pkg/trace/trace.go

```diff
@@ -323,3 +323,12 @@ func FromHeaders(ctx context.Context, hdrs map[string][]string, name string) con
 	}
 	return defaultTracer.fromHeaders(ctx, hdrs, name)
 }
+
+// IsSampled returns true if the given trace ID is sampled based on the current sampling rate.
+// This API is mostly for remote trace ID correlation.
+func IsSampled(traceID string) bool {
```

Comment:
Just curious: do we need `traceID` as a param here?  Wouldn't it be better to make this API take the `context` and identify if the current trace is being sampled or not?

Reply:
We have the trace.ID()` function which returns the trace ID out of the context ... so this part is covered.

As for the IsSampledWithContext - my wild guess is it is not needed because it can be easily achieved with `trace.IsSampled(trace.ID())`

I do not have a strong opinion around either using the context (and hiding trace ID as impl) - but given that we already have the trace.ID function to expose raw traceID out it would be awkward to ask people to wrap the trace ID back into the context in order to get the IsSampled method called on it (in case they already pass the ID as a string around)

Reply:
If we support enabling/disabling sampling or changing sampling rate based on stuff in the context, an API that is based on context is likely to be more accurate.

Not disputing whether we need the traceID based API as well but I think if the intention is to figure if the current trace is sampled, it would be better to model that API via the context.

Reply:
The intention of this API is to figure out if the remote trace ID is sampled, not the local one, so that it could be 'safely linked' by the recipient of the trace.

It is very hard to think of a case where somebody would need IsSampled equivalent for the local traces. If yes, then we can always add dedicated IsSampledFromContext that will take care of trace ID + whatever local overrides where applied to force the sample.

Reply:
An alternate design is to model it like so:

```golang
    trace.CurrentSampler(ctx).IsSampled(trace.ID(ctx)) ...
```

It does require a lot more changes but it is probably more elegant as we can make the `trace.CurrentSampler` return the default sampler for the offline cases.

Reply:
If the API is explicitly for remote traces, how would we deal with sample rates and such?

It seems wrong to use the current sampling configuration for remote traces and so I'd suggest we do something like `trace.Sampler(config/sample_rate).IsSampled(traceID)` for that case rather than `trace.IsSampled`.

Reply:
Actually, Link does not expose raw Context object - simply because it is a Link and we do not want to accidentally generate context out of the link. That means the link itself should be the one to tell if it is sampled or not.

Maybe IsSampled should be moved into the Link struct instead to clearly scope it to the link only?

Reply:
I moved it to the Link struct to scope it down, and renamed into IsSampledIfLocal + doc about heuristics and reasoning around this. LMK if it looks better.

Reply:
Actually there is another concern here. Although we are sampling at the same rate everywhere today, that will not always be the case. For example is service A is sampling at 10% and service B is sampling 1%, then service B would still have a lot of misses for the traces from service A.

To solve this, I think we have two options:

- We need to pass through the original sampling percent and use that to create a new sampler at the same percentile. This seems like a lot to do in order to link the traces.
- We can pass through a force header only on the traces that get sampled from A. That would mean that regardless of the sampling rate of B, we will force B to sample that trace.

Reply:
this is indeed a better fix, but I propose we do it incrementally because it wont break API. Without the current fix e2e trace feature is completely unusable - so with this fix we will at least have the MOST COMMON case covered in which sampling between producer and consumer are same. With your fix it will be even more accurate indeed, but it requires three more iterations that IMHO can happen later too.

Reply:
The earlier suggestion I had (see below) forces the sample rate to be explicit -- this will atleast make it clear that code built with it is going to have all these problems.   I'd personally prefer we do the force-trace approach that @artemchernyak-outreach suggests though -- better to have things be robust when it comes to observability.

> It seems wrong to use the current sampling configuration for remote traces and so I'd suggest we do something like trace.DefaultSampler(config/sample_rate).IsSampled(traceID) for that case rather than trace.IsSampled.

Reply:
@vk-outreach clerk library does not have access to the inner sample rate - it is deep inside the trace configuration. 

If you would still prefer to expose the Sampler, maybe it should be trace.CurrentSampler() instead which uses the hidden sample rate instead of forcing callers to magically  get the sample rate somehow.